### PR TITLE
Add Typescript typings

### PR DESCRIPTION
**Pull request by revskill10**
**Description**
Add Typescript typings

### CODE REVIEW

**Comment by nithincvpoyyil** [comment]
Hi @Aljullu  - just being curious, any plan to merge this PR?

**Comment by Aljullu** [comment]
Thanks for the PR @revskill10 and my apologizes for the delay answering. The fact is that my knowledge of TypeScript is almost 0, so I'm not sure what are the pros/cons of including this in the default package. I was planning to take a deep look at this PR at some point, but I have limited time to spend in this project, so I have been delaying it more than I initially thought.

With a quick look at the code, one of the questions I get is whether we should also include `interface LazyLoadComponentProps` in addition to `interface LazyLoadImageProps`.

cc @nithincvpoyyil

**Comment by rodrigomf24** [comment]
It would be great to have support for Typescript ❤️  I am willing to help

**Comment by nithincvpoyyil** [comment]
> Thanks for the PR @revskill10 and my apologizes for the delay answering. The fact is that my knowledge of TypeScript is almost 0, so I'm not sure what are the pros/cons of including this in the default package. I was planning to take a deep look at this PR at some point, but I have limited time to spend in this project, so I have been delaying it more than I initially thought.
> 
> With a quick look at the code, one of the questions I get is whether we should also include `interface LazyLoadComponentProps` in addition to `interface LazyLoadImageProps`.
> 
> cc @nithincvpoyyil

@Aljullu : Yes, its better to give a one shot for all.  
@revskill10 : Could you please update that as well, in this PR?

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    threshold?: number;
+    visibleByDefault?: boolean;
+    wrapperClassName?: string;
+    src: string | null;
+    width?: number;
+    height?: number;
+    alt?: string | null;
+  }
+
+  export const LazyLoadImage: (props: LazyLoadImageProps) => JSX.Element;
```
Comment (nithincvpoyyil) [inline index.d.ts]:
```
// LazyLoadComponent defs
interface  LazyLoadComponentProps {
  afterLoad:;()=>void;
  beforeLoad: ()=>void;
  useIntersectionObserver: boolean;
  visibleByDefault: boolean;
};

 export const LazyLoadComponent: (props: LazyLoadComponentProps ) => JSX.Element;

// trackWindowScroll  HOC 
 export function trackWindowScroll<C extends React.ComponentType<any>>(component: C): C;
```

@revskill10 , @Aljullu

**Comment by danvk** [comment]
@nithincvpoyyil @revskill10 @Aljullu FYI another option for TypeScript declarations is to host them separately on [DefinitelyTyped][1]. There are pros & cons to bundling the typings with this repo vs. externally, but if you're not familiar w/ TS and find the prospect of maintaining type declarations daunting, it might be a good option.

If you'd prefer that approach, I'm happy to create a PR on DT and get it through.

Thanks for the great library!

[1]: https://github.com/DefinitelyTyped/DefinitelyTyped

**Comment by Aljullu** [comment]
> If you'd prefer that approach, I'm happy to create a PR on DT and get it through.

That would be great @danvk! As you said, I'm a bit reluctant to add TypeScript declarations in this repo because I won't be able to maintain them and wouldn't like to merge something that might block future releases, so hosting them in DefinitelyTyped sounds like a good solution to me.

Let me know if I can help in any way!

**Comment by danvk** [comment]
Here it is: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40565

@revskill10 if you could look at the type declarations, that'd be great. I've listed you as an author / maintainer.

@Aljullu I've tried to figure out expected usage for things like `trackWindowScroll` but I'm not entirely sure if I got it right. It would be great if you could look at the tests.

**Comment by danvk** [comment]
... and they're merged. You can get type declarations by running:

    npm install -D @types/react-lazy-load-image-component

**Comment by Aljullu** [comment]
Thank you @danvk! I added a link in the README so everybody can find them:

https://github.com/Aljullu/react-lazy-load-image-component#features

**Comment by nithincvpoyyil** [comment]
@danvk :+1:

**Pull request closed without merging**